android/app/src/main/java/app/organicmaps/routing/NavigationService.java

```diff
@@ -210,16 +210,21 @@ public void onLocationUpdated(@NonNull Location location)
     if (!routingController.isNavigating())
       return;
 
+    final String[] turnNotifications = Framework.nativeGenerateNotifications();
+    if (turnNotifications != null)
+      TtsPlayer.INSTANCE.playTurnNotifications(getApplicationContext(), turnNotifications);
+
     // TODO: consider to create callback mechanism to transfer 'ROUTE_IS_FINISHED' event from
     // the core to the platform code (https://github.com/organicmaps/organicmaps/issues/3589),
     // because calling the native method 'nativeIsRouteFinished'
     // too often can result in poor UI performance.
+    // This check should be done after playTurnNotifications() to play the last turn notification.
     if (Framework.nativeIsRouteFinished())
+    {
       routingController.cancel();
-
-    final String[] turnNotifications = Framework.nativeGenerateNotifications();
-    if (turnNotifications != null)
-      TtsPlayer.INSTANCE.playTurnNotifications(getApplicationContext(), turnNotifications);
+      stopSelf();
+      return;
```

Comment:
The previous code did not return early, but did something below. Are you sure that it's not needed in this case?

Reply:
The code below updates the notification, which is not really useful because it will be removed on the next event loop when the service is stopped after stopSelf() call. I can remove `return` from here, it changes nothing.

Reply:
Maybe only missing "beep" for a camera can be a concern...